src/expint.jl

```diff
@@ -380,22 +386,24 @@ function expint(ν::Number, z::Number, niter::Int=1000)
     if abs(ν) > 50 && !(isreal(ν) && real(ν) > 0)
         throw(ArgumentError("Unsupported order |ν| > 50 off the positive real axis"))
     end
-    ν, z = promote(ν, float(z))
-    if typeof(z) <: Real && z < 0
+
+    z, = promote(float(z), ν)
+    if isnan(ν) || isnan(z)
+        return oftype(z, NaN)
```

Comment:
Maybe `oftype(z, NaN) * z` to make it return `NaN + NaN*im` in the complex case.